File: behavioral/typescript/iterator_structured/book_collection.ts
import { Iterator, Aggregate } from './interfaces';

export class BookCollection implements Aggregate<string> {
  private items: string[] = [];
  
  add(item: string): void {
    this.items.push(item);
  }
  
  createIterator(): Iterator<string> {
    return new BookIterator(this);
  }
  
  size(): number {
    return this.items.length;
  }
  
  get(index: number): string {
    return this.items[index];
  }
}

export class BookIterator implements Iterator<string> {
  private collection: BookCollection;
  private position: number = 0;
  
  constructor(collection: BookCollection) {
    this.collection = collection;
  }
  
  hasNext(): boolean {
    return this.position < this.collection.size();
  }
  
  next(): string {
    if (this.hasNext()) {
      return this.collection.get(this.position++);
    }
    return '';
  }
}

File: behavioral/dart/iterator_structured/book_collection.dart
import 'interfaces.dart';

class BookCollection implements Aggregate<String> {
  final List<String> _items = [];
  
  void add(String item) {
    _items.add(item);
  }
  
  @override
  Iterator<String> createIterator() {
    return BookIterator(this);
  }
  
  int size() => _items.length;
  
  String get(int index) => _items[index];
}

class BookIterator implements Iterator<String> {
  final BookCollection _collection;
  int _position = 0;
  
  BookIterator(this._collection);
  
  @override
  bool hasNext() {
    return _position < _collection.size();
  }
  
  @override
  String next() {
    if (hasNext()) {
      return _collection.get(_position++);
    }
    return '';
  }
}
